platforms/core-runtime/gradle-cli-main/src/main/java/org/gradle/launcher/GradleMain.java

```diff
@@ -27,11 +28,11 @@ public class GradleMain {
     public static void main(String[] args) throws Exception {
         int currentMajorJavaVersion = JavaVersionParser.parseCurrentMajorVersion();
         if (currentMajorJavaVersion < SupportedJavaVersions.MINIMUM_CLIENT_JAVA_VERSION) {
-            System.err.printf(
-                "Gradle requires JVM %d or later to run. You are currently using JVM %d.%n",
+            System.err.printf(UnsupportedJavaRuntimeException.getIncompatibleJavaVersionForProcessErrorMessage(
```

Comment:
🤔 What is the benefit to our users that we tell them the `The Gradle CLI Client` is compatible with Java 8, only to tell them later that they need Java 17?

Should we prefer to communicate to our users that they need Java 17 here already? That we compile this part down to Java 8 is an implementation detail IMO.

Reply:
Running the client on 8 and the daemon on 17 (via daemon toolchains for example) is supported. If the user runs the client on an unsupported version (7 or 8) then telling them they need to run with 17 isn't really true

Reply:
Running the client on 8 and the daemon on 17 (via daemon toolchains for example) is supported. If the user runs the client on an unsupported version (7 or 8) then telling them they need to run with 17 isn't really true

Reply:
In any case, if we go through with compiling this code to Java 8, this will fail with an incompatible class version exception, and we can just remove this code. 

But for the time being, this is IMO more correct. 

This was more a side effect of the changes I intended to write -- which was removing the `Pattern` from test expectations, in the cases where we weren't really sure what the test output should be. Now the test expectations are more certain

Reply:
If we only communicate to the user that they need Java 8 here, then that forces us to keep an entrypoint in the Gradle Daemon that is compatible with Java 8 to return an error communicating to the user they need Java 17 (or an even more recent version if daemon toolchains increases that further)

If, on the other hand, we inform the user here that at least Java 17 is required, we can completely remove any code < Java 17 from the Gradle Daemon and don't add the complexity for producing multiple bytecode levels there.

Reply:
> that forces us to keep an entrypoint in the Gradle Daemon that is compatible with Java 8 to return an error communicating to the user they need Java 17

I don't think this is a requirement. 

The client currently verifies that the JDK it is about to use to start a daemon is compatible with 17. That is what the `assertIsSupportedDaemonJvmVersion` call in `DefaultDaemonStarter` is doing. This is the message that contains "Your build is currently configured". This is the message that most of our tests currently emit